apps/settings/js/panels/simpin_dialog/simpin_dialog.js

```diff
@@ -0,0 +1,590 @@
+          retryCount: req.error.retryCount,
+          defer: defer
+        });
+      };
+
+      return defer.promise;
+    },
+
+    /**
+     * this function is used when we are going to update fdn contacts under
+     * call panel.
+     *
+     * @memberOf SimPinDialog
+     * @access private
+     * @return {Promise}
+     */
+    _updateFdnContact: function() {
+      //  Updates a FDN contact. For some reason, `icc.updateContact` requires
+      //  the  pin input value instead of delegating to `icc.setCardLock`.
+      //  That means that, in case of failure, the error is different that the
+      //  one that `icc.setCardLock` gives. This means that we have to handle
+      //  it separatedly instead of being able to use the existing
+      //  `handleCardLockError` above.
+      //
+      //  Among other things, it doesn't include the retryCount, so we can't
+      //  tell the user how many remaining tries she has. What a mess.
+      //  This should be solved when bug 1070941 is fixed.
+      var promise = new Promise((resolve, reject) => {
+        var fdnContactInfo = this._pinOptions.fdnContactInfo;
+        var req = this._icc.updateContact('fdn',
+          fdnContactInfo, this._elements.pinInput.value);
+
+        req.onsuccess = () => {
+          resolve(fdnContactInfo);
+        };
+
+        req.onerror = (e) => {
+          switch (req.error.name) {
+            case 'IncorrectPassword':
+              // TODO: count retries (not supported by the platform) ->
+              // Bug 1070941
+              this._initUI('get_pin2');
+              this._showMessage('fdnErrorMsg');
+              this._elements.pinInput.value = '';
+              this._elements.pinInput.focus();
+              reject();
+              break;
+
+            case 'SimPin2':
+              // TODO: count retries (not supported by the platform) ->
+              // Bug 1070941
+              this._initUI('get_pin2');
+              this._showMessage('fdnErrorMsg');
+              this._elements.pinInput.value = '';
+              this._elements.pinInput.focus();
+              reject();
+              break;
+
+            case 'SimPuk2':
+              this._initUI('unlock_puk2');
+              this._elements.pukInput.focus();
+              reject();
+              break;
+
+            case 'NoFreeRecordFound':
+              alert(l10n.get('fdnNoFDNFreeRecord'));
+              resolve(fdnContactInfo);
+              break;
+
+            default:
+              resolve(fdnContactInfo);
+              throw new Error('Could not edit FDN contact on SIM card', e);
+          }
+        };
+      });
+      return promise;
+    },
+
+    /**
+     * We will try to handle all returned errors here when requests are
+     * returned from icc.
+     *
+     * @memberOf SimPinDialog
+     * @param {Object} options
+     * @access private
+     * @return {Promise}
+     */
+    _handleCardLockError: function(options) {
```

Comment:
Passing the defer object to this function makes it depending on something that has nothing to do with the function. I would suggest to make this a simple async function that returns a promise.